src/Sarif/Visitors/LogicalLocationIndexRemapper.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    public class LogicalLocationIndexRemapper : SarifRewritingVisitor
+    {
+        private IList<LogicalLocation> _previousLogicalLocations;
+
+        public LogicalLocationIndexRemapper(IList<LogicalLocation> previousLogicalLocations, IDictionary<LogicalLocation, int> remappedLogicalLocations)
+        {
+            _previousLogicalLocations = previousLogicalLocations;
+            RemappedLogicalLocations = remappedLogicalLocations;
+        }
+
+        public IDictionary<LogicalLocation, int> RemappedLogicalLocations { get; private set; }
+
+        public override Location VisitLocation(Location node)
+        {
+            if (_previousLogicalLocations == null) { return node; }
```

Comment:
>_previousLogicalLocations [](start = 16, length = 25)

If `_previoiusLogicalLocations` is `null`, there's no point visiting any of the locations. So you could cut the visit off early, like:
```
public override Result VisitResult(Result node)
{
    if (_previousLogicalLocations == null) { return node; } // ... and remove this line from VisitLocation.

    base.VisitResult(node);
}
```